### DISCUSSION TITLE
Custom background to TextBox

### DISCUSSION THREAD

**Discussion by iksi4prs** [AvaloniaUI/Avalonia#11740] [Q&A]
Hi,
I'm trying to create custom background for TextBox.
My target is to have different background color per line, so I thought about implement some custom brush.
But even before, if I just change Background in xaml/code to SolidColorBrush, it doesn't work, and always stays "White".
If I change in xaml, it will work in the designer, but when I run the app, it will change to white.
Looks like "someone" is overriding my code.
Any idea what to fix, or to look for ?
Thanks.

**Comment by timunie** [comment]
Hi @iksi4prs 

first of all, let me know which Avalonia version you use. 

In general I can give these hints for customizing styles:
- Use DevTools [F12] to analyze current values https://docs.avaloniaui.net/docs/getting-started/developer-tools
- Lookup original styles for the Avalonia version you use 
- Make sure to read style docs, esp. troubleshooting https://docs.avaloniaui.net/docs/styling/troubleshooting
- ProTip: If you need alternating style, check nth-child-selector: https://docs.avaloniaui.net/docs/styling/selectors#nth-child

**Reply by iksi4prs** [reply]
@timunie 

Thanks for your help.
I'm using version 11.0.0-preview5.

As you've suggested, I checked the style using the DevTools (F12).

In my code, in XAML, I add the following selector/setter:
```
<Style Selector="TextBox:focus-within">
    <Setter Property="FontSize" Value="30"/>
    <Setter Property="Background">
        <SolidColorBrush Color="Orange"/>
    </Setter>
</Style>
``` 

When I check the and move focus to the TextBox, I clearly see the font changes to correct size, (to the bigger "30"),
but color stays white, and doesn't change to orange.

When I check in the DevTools, as you can see in attached image, the computed/active value should the be orange color,
(maybe I'm wrong?)
but its not the color used.
Any suggestion ?


![image](https://github.com/AvaloniaUI/Avalonia/assets/65832579/5e8f53e4-2e0d-4d3c-81d1-b30f7796756c)

**Reply by timunie** [reply]
Most likely the Background is set locally via dynamic resource. In that case, a style cannot override it. But you can simply replace the given Brush with whatever you need.

**Reply by timunie** [reply]
### Alternative 1: Use the same selector as in original style
```xml
<Style Selector="TextBox:focus">
	<Setter Property="FontSize" Value="30"/>
</Style>

<Style Selector="TextBox:focus /template/ Border#PART_BorderElement">
	<Setter Property="Background">
		<SolidColorBrush Color="Orange"/>
	</Setter>
</Style>
```

Source: https://github.com/AvaloniaUI/Avalonia/blob/d78dc2aa3bcaa0be2ddca6c5d5627de4dcd79e9b/src/Avalonia.Themes.Fluent/Controls/TextBox.xaml#L207-L220

### Alternative 2: Override the needed resource (can be done in Style, any Node, but also in App.Resources)
```xml
<UserControl.Resources>
    <SolidColorBrush x:Key="TextControlBackgroundFocused">Orange</SolidColorBrush>
</UserControl.Resources>
<UserControl.Styles>
    <Style Selector="TextBox:focus">
        <Setter Property="FontSize" Value="30" />
    </Style>
</UserControl.Styles>
```

Resource found here: https://github.com/AvaloniaUI/Avalonia/blob/d78dc2aa3bcaa0be2ddca6c5d5627de4dcd79e9b/src/Avalonia.Themes.Fluent/Accents/FluentControlResources.xaml#L235
